File: base/src/main/java/me/shetj/base/tools/time/CodeUtil.java
package me.shetj.base.tools.time;

import android.app.Activity;
import androidx.annotation.Keep;
import android.widget.TextView;

import java.util.Timer;
import java.util.TimerTask;

@Keep
public class CodeUtil {
	private TextView codeTV;
	private Timer timer;
	private TimerTask timerTask;
	private Activity activity;
	private int second;
	private static final int DEF_TIME_LENGTH = 60;
	private int timeLength = DEF_TIME_LENGTH;

	public CodeUtil(Activity activity, TextView tv) {
		codeTV = tv;
		this.activity = activity;
	}

	public CodeUtil(Activity activity, TextView tv, int timeLength) {
		codeTV = tv;
		this.activity = activity;
		this.timeLength = timeLength;
	}

	public void start() {
		stopTimer();
		second = timeLength;
		codeTV.post(new Runnable() {
			@Override
			public void run() {
				codeTV.setEnabled(false);
			}
		});

		timer = new Timer();
		timerTask = new CodeTimerTask();
		timer.scheduleAtFixedRate(timerTask, 0, 1000);
	}

	public int getSecond(){
		return second;
	}

	private void stopSendCode() {
		stopTimer();
		codeTV.post(new Runnable() {
			@Override
			public void run() {
				codeTV.setText("获取验证码");
				codeTV.setEnabled(true);
			}
		});
	}

	private void stopTimer() {
		if (timer != null) {
			timer.cancel();
			timer.purge();
			timer = null;
		}
		if (timerTask != null) {
			timerTask.cancel();
			timerTask = null;
		}
	}

	public void stop() {
		second = 0;
	}

	class CodeTimerTask extends TimerTask {

		@Override
		public void run() {
			if (second <= 0) {
				stop();
			} else {
				onWaiting();
			}
		}

		private void stop() {
			stopSendCode();
		}

		private void onWaiting() {
			codeTV.post(new Runnable() {
				@Override
				public void run() {
					codeTV.setText(String.format("%s秒后重发", second));
					second--;
				}
			});
		}
	}
}

File: base/src/main/java/me/shetj/base/tools/time/CodeUtil.kt
package me.shetj.base.tools.time

import android.widget.TextView
import androidx.annotation.Keep
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.core.Flowable
import io.reactivex.rxjava3.disposables.CompositeDisposable
import io.reactivex.rxjava3.disposables.Disposable
import timber.log.Timber
import java.util.concurrent.TimeUnit

@Keep
class CodeUtil {
    private var codeTV: TextView? = null
    private var second: Int = 0
    private var timeLength = 60
    private var mCompositeDisposable: CompositeDisposable? = null

    constructor(tv: TextView) {
        codeTV = tv
    }

    constructor(tv: TextView, timeLength: Int) {
        codeTV = tv
        this.timeLength = timeLength
    }

    fun start() {
        second = timeLength
        codeTV?.post { codeTV?.isEnabled = false }
        val subscribe =
                Flowable.interval(1, TimeUnit.SECONDS)
                        .take(timeLength.toLong())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe({
                            second--
                            codeTV?.text = String.format("%s秒后重发", second)
                        }, {
                            stop()
                        }, {
                            codeTV?.text = "获取验证码"
                            codeTV?.isEnabled = true
                        })
        addDispose(subscribe)
    }


    fun stop() {
        Timber.i("stop_code_util")
        unDispose()
        this.mCompositeDisposable = null
        codeTV?.post {
            codeTV?.text = "获取验证码"
            codeTV?.isEnabled = true
        }
    }

    /**
     * 将 [Disposable] 添加到 [CompositeDisposable] 中统一管理
     * 可在 [android.app.Activity.onDestroy]中使用
     */
    private fun addDispose(disposable: Disposable) {
        if (mCompositeDisposable == null) {
            mCompositeDisposable = CompositeDisposable()
        }
        mCompositeDisposable?.add(disposable)
    }

    /**
     * 停止集合中正在执行的 [RxJava] 任务
     */
    private fun unDispose() {
        if (mCompositeDisposable != null) {
            mCompositeDisposable?.clear()
        }
    }

}

